Add Spring WebFlux and WebClient tracing.

csabakos opened the pull request:
This adds Spring WebFlux and WebClient tracing.

I will also add auto-configuration for https://github.com/opentracing-contrib/java-reactor to `opentracing-spring-cloud`, which will complement these changes with automatic activation of the span in handler functions.

csabakos commented on opentracing-spring-web/pom.xml:
```diff
@@ -37,5 +42,17 @@
     </dependency>
+    <dependency>
+      <groupId>com.github.tomakehurst</groupId>
+      <artifactId>wiremock-jre8</artifactId>
```
I had to replace `MockRestServiceServer` in the tests with something else (WireMock), because `MockRestServiceServer` does not support `WebClient`.

csabakos commented on pom.xml:
```diff
@@ -35,8 +35,8 @@
   <properties>
-    <maven.compiler.source>1.7</maven.compiler.source>
-    <maven.compiler.target>1.7</maven.compiler.target>
+    <maven.compiler.source>1.8</maven.compiler.source>
```
Spring Framework 5 requires JRE 8 so I believe this is safe to change.

pavolloffay replied:
Note that this is not only SB, the web package can be used with old spring web apps.

geoand replied:
@pavolloffay Spring Framework 5 + required JDK 8+ see [here](https://github.com/spring-projects/spring-framework/wiki/Upgrading-to-Spring-Framework-5.x#upgrading-to-version-50)

pavolloffay commented on README.md:
```diff
@@ -2,22 +2,32 @@
 get automatically traced different set of technologies e.g. `spring-cloud-netflix`, JMS or even more then
 use project [opentracing-spring-cloud](https://github.com/opentracing-contrib/java-spring-cloud) instead.
 
+For reactive applications, it is especially recommended to use `reactor` tracing from
```
Is reactor supported in opentracing-spring-cloud?

csabakos replied:
I'm planning to add that in a separate pull request. I already have the code ready for it, but `opentracing-reactor` needs to be published to Maven central first. (See https://github.com/opentracing-contrib/java-reactor/issues/1)

pavolloffay commented on opentracing-spring-web-itest/webflux/src/test/java/io/opentracing/contrib/spring/webflux/webfilter/itest/webflux/WebFluxJettyMinimalIntegrationTest.java:
```diff
@@ -0,0 +1,112 @@
+        Assert.assertEquals("GET", span.tags().get(Tags.HTTP_METHOD.getKey()));
+        Assert.assertEquals(TEST_REST_TEMPLATE.getRootUri() + "/", span.tags().get(Tags.HTTP_URL.getKey()));
+        Assert.assertEquals(200, span.tags().get(Tags.HTTP_STATUS.getKey()));
+        Assert.assertEquals("WebFlux/java-spring-web", span.tags().get(Tags.COMPONENT.getKey()));
```
Change `java-spring-webflux` to be consistent with other component names in this repo?

pavolloffay commented on opentracing-spring-web-starter/pom.xml:
```diff
@@ -24,6 +24,11 @@
+    <dependency>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-webflux</artifactId>
+      <scope>provided</scope>
```
I think this should be optional? We don't want users to depend on webflux

csabakos replied:
Thanks, `optional` makes more sense than `provided`.

pavolloffay commented on opentracing-spring-web-starter/src/main/java/io/opentracing/contrib/spring/web/starter/WebClientTracingAutoConfiguration.java:
```diff
@@ -0,0 +1,64 @@
+ */
+@Configuration
+@ConditionalOnBean(Tracer.class)
+@ConditionalOnClass(WebClient.class)
```
Should we also do conditional on reactive?

pavolloffay replied:
@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.REACTIVE)

csabakos replied:
`@ConditionalOnWebApplication` refers to the server technology.

One could use the reactive `WebClient` in a Servlet-based server application, so I don't think that adding that conditional would be appropriate.

pavolloffay commented on opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/WebClientSpanDecorator.java:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        @Override
+        public void onCancel(final ClientRequest httpRequest, final Span span) {
```
When does cancel happen?

csabakos replied:
Excellent question! It looks like cancellation can happen in various circumstances:
- If the user of the WebClient "unsubscribes" from the request (for example because they have a timeout that they reached and they no longer wish to wait for the result). I think in this case I think it is correct to log an error on the span, since timeout can be considered an error.
- In various Reactor constructs, `cancel()` is called even when the `Mono` (successfully) emits an item. An example of this is [`Flux#next()`](https://github.com/reactor/reactor-core/blob/7bf86a81734bf1dc1a468a99dfdad2590c089de9/reactor-core/src/main/java/reactor/core/publisher/MonoNext.java#L75). Since this is part of the normal (successful) behavior, I don't think we should log an error.

There does not seem to be a good way to differentiate between the above two cases in the tracing code, so I changed this to log a `cancelled` event instead of an error. It will be up to the user who is looking at the trace logs to determine if the cancellation indicates an error or not. (For example the user can infer this from the HTTP status code, etc.)

pavolloffay replied:
If it is a part of normal execution flow and it happes often we can avoid adding meta-data at all.

csabakos replied:
As an application developer, it is still valuable information that the subscription was cancelled. Cancellation does not happen in simple WebClient flows, only if you're doing something more complex (like retries, or firing off multiple requests and waiting for the fastest one to return).

I think it's worth have these events in the trace logs.

If other people find them bothersome, they can file an issue and we can make it optional, but I think that won't be the case.

csabakos commented:
> @csabakos thanks for the PR! It looks good.

@pavolloffay Thank you for the review!

> I would like to see a test where a client is used to from multiple threads (assuming it supports concurrent processing) and creates requests to instrumented web endpoint (can be reactive too)
> 
> [java-spring-web/opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java](https://github.com/opentracing-contrib/java-spring-web/blob/5d7207a43ec4b8efd0314fb42ae8810eed43f4fa/opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java#L65)
> 
> Line 65 in [5d7207a](/opentracing-contrib/java-spring-web/commit/5d7207a43ec4b8efd0314fb42ae8810eed43f4fa)
> 
>  public void testMultipleRequests() throws InterruptedException, ExecutionException {

I implemented the exact test that you linked to for WebClient: [opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingWebClientTest.java](https://github.com/csabakos/java-spring-web/blob/e41a79f48c626f48e200459dc7f4c61e4cdeb051/opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingWebClientTest.java#L68). Is that sufficient or did you have something else in mind?

> I have noticed that some files were taken from spring-cloud-sleuth. I think we should add their license (We already doing it for some files in spring-cloud repository)

Conveniently their license is also Apache License 2.0, but I will gladly add it. Would you mind pointing me to an existing example of what exactly needs to be done?

pavolloffay commented:
> Conveniently their license is also Apache License 2.0, but I will gladly add it. Would you mind pointing me to an existing example of what exactly needs to be done?

I think just add the license to the files and then exclude them in pom https://github.com/opentracing-contrib/java-spring-cloud/blob/master/pom.xml#L380

pavolloffay commented on opentracing-spring-web-starter/src/main/java/io/opentracing/contrib/spring/web/starter/WebFluxTracingAutoConfiguration.java:
```diff
@@ -0,0 +1,73 @@
+@AutoConfigureAfter(TracerAutoConfiguration.class)
+@EnableConfigurationProperties(WebTracingProperties.class)
+@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.REACTIVE)
+@ConditionalOnProperty(name = "opentracing.spring.web.enabled", havingValue = "true", matchIfMissing = true)
```
`CONFIGURATION_PREFIX` can be used

pavolloffay commented on opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingWebClientTest.java:
```diff
@@ -0,0 +1,133 @@
+            final String requestUrl = url + i;
+
+            final Scope parentSpan = mockTracer.buildSpan("foo").startActive(false);
+            parentSpan.span().setTag("request-url", requestUrl);
```
Add tag on spanBuilder

pavolloffay commented on opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingWebClientTest.java:
```diff
@@ -0,0 +1,133 @@
+        for (int i = 0; i < numberOfCalls; i++) {
+            final String requestUrl = url + i;
+
+            final Scope parentSpan = mockTracer.buildSpan("foo").startActive(false);
```
Create a span directly and remove `cont`. cont does not seems to be needed

geoand commented:
> @geoand could you please review too?

I will do that tomorrow :)

csabakos commented:
> > Conveniently their license is also Apache License 2.0, but I will gladly add it. Would you mind pointing me to an existing example of what exactly needs to be done?
> 
> I think just add the license to the files and then exclude them in pom https://github.com/opentracing-contrib/java-spring-cloud/blob/master/pom.xml#L380

Done. The pom did not have `license-maven-plugin` so I went ahead and added that and updated all of the license headers (so that the build passes).

geoand commented:
@csabakos Can you please resolve the conflict? Thanks

csabakos commented:
> @csabakos Can you please resolve the conflict? Thanks

@geoand resolved by merging `master` into my feature branch. (Let me know if that's not the preferred way to do this.)

geoand commented:
That's up to @pavollofay to decide 😎

geoand commented on opentracing-spring-web-starter/src/main/java/io/opentracing/contrib/spring/web/starter/WebClientTracingAutoConfiguration.java:
```diff
@@ -0,0 +1,64 @@
+@ConditionalOnClass(WebClient.class)
+@ConditionalOnProperty(prefix = WebClientTracingProperties.CONFIGURATION_PREFIX, name = "enabled", matchIfMissing = true)
+@AutoConfigureAfter(TracerAutoConfiguration.class)
+@EnableConfigurationProperties(WebClientTracingProperties.class)
```
Should `WebClientTracingProperties` be renamed? I mean the name is a little misleading since it's configuration for `RestTemplate`, `AsyncRestTemplate` and `WebClient`.
Also, the comments in `WebClientTracingProperties` should be updated to reflect this fact

csabakos replied:
I don't have an opinion about renaming - I could see arguments both pro and con. I'll leave the decision up to you and @pavolloffay. (Because the class is `public`, it is in theory part of the public API, so that'd be one reason for not renaming it.)

I will update the comments though.

geoand replied:
It's true that it's public, public configuration classes would rarely be used directly in application code. I would assume that at least a minor version bump would be warranted even without the rename, so some minor breaking could probably be expected

geoand commented on opentracing-spring-web-starter/pom.xml:
```diff
@@ -24,6 +43,12 @@
+    <dependency>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-webflux</artifactId>
+      <scope>compile</scope>
```
This is default scope, so perhaps best to omit it?

pavolloffay replied:
Shouldn't this be optional?

csabakos replied:
Nice catch! Looks like I only updated the POM in `opentracing-spring-web-starter` but not in `opentracing-spring-web`. Fixed now.

geoand commented:
Looks good from a Spring perspective :+1:

pavolloffay commented:
@geoand can this be merged?

geoand commented:
@pavolloffay yeah it LGTM. Should I merge and cut a release?

pavolloffay commented:
:+1:

geoand commented:
Thanks for the awesome work @csabakos!

csabakos commented:
Thanks y'all for the prompt review and good comments!

I'll work on adding the Spring Boot starter for `opentracing-reactor` to `opentracing-spring-cloud` after https://github.com/opentracing-contrib/java-reactor/issues/1 is completed.

geoand commented:
Thanks again @csabakos!

ghilainm commented:
@csabakos Seems like the filter is not instantiated if spring-mvc is not on the classpath. Is it intented? I am using spring cloud gateway and the filter is not created because of this :).
